### PR TITLE
RaiseAsync

### PR DESCRIPTION
**Pull request by dvorn**
PR for #678

### CODE REVIEW

**Comment by wldevries** [comment]
Looks good, but I think default parameters is something that should not be done in libraries. We should choose a default behavior and add a second method that gives the choice.

**Comment by dvorn** [comment]
@wldevries You're right.  FxCop always warns about that.  The reason is that other .NET languages (not C#) may not have a concept of default parameters.

**Comment by dvorn** [comment]
@wldevries Overload added.

File: Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs [hunk 1]
```diff
@@ -48,6 +48,19 @@ public void Raise(T context, Action<T> callback)
...
+        }
+
+        /// <summary>
+        /// Fires the Raised event asynchronously. Please note that this request may never return
+        /// if the InteractionRequest is unhandled.
+        /// </summary>
+        /// <param name="context">The context for the interaction request.</param>
         /// <param name="executeSynchronously">
         /// Has no effect if the popup window is not modal.
         /// If the popup window is modal and executeSynchronously is true, retruns already completed task.
```
Comment (joacar) [inline Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs]: Just a minor spelling issue: retruns should be returns

**Comment by dvorn** [comment]
@joacar Thanks!

**Comment by brianlagunas** [comment]
Why exactly does an  Async method have an option to run sync (executeSynchronously)?  Wouldn't devs just use the normal Raise method instead?

**Comment by brianlagunas** [comment]
I'm concerned about the addition of this property to support a very specific use case.  My gut tells me this type of behavior should be handled in application code, not in the Prism API.  In application code you would determine if you should call Raise or RaiseAsync.  It feels wrong to have an async method have a parameter to execute it sync just to support a single use case.

File: Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs [hunk 2]
```diff
@@ -47,10 +48,59 @@ public void Raise(T context, Action<T> callback)
...
+            {
+                return this.RaiseImplAsync(context);
+            }
+        }
+
+        private async Task<T> RaiseImplAsync(T context)
+        {
+            // SynchronizationContext exists in actual program running on GUI Dispatcher thread.  Make sure
+            // it exists in tests.
+            if (SynchronizationContext.Current == null)
```
Comment (brianlagunas) [inline Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs]: is this here soley for making sure tests have a sync context? If so, this does not belong here.  You setup your sets to make sure it has a sync context.

**Comment by dansiegel** [comment]
This seems a very wrong way of exposing the functionality. I understand there are lots of use cases where there is a need to execute syncronosly, however the better way of handling this is to not have a bool in an Async method and rather do something like:

```cs
public Task<T> RaiseAsync<T>(T context)
public T Raise<T>(T context)
```

**Comment by dvorn** [comment]
@dansiegel Then you end up in writing duplicated code - one code for sync version, another for async.  And the sync version is not testable.

**Comment by brianlagunas** [comment]
@dvorn If your app requires this, then you write your helper method to perform the action you require and you don't duplicate code in your app.  But this is an app requirement and does not belong in the Prism API

**Comment by dvorn** [comment]
@brianlagunas I already have this in my application.  The point is to have this functionality available to all developers using Prism.  If you continue your reasoning, you end up with the Prism is not needed at all - developers can write all the code by themselves.  What is the purpose of having a library?  They may invent and code it by themselves.

File: Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs (hunk 2 above)
Comment (dvorn) [inline Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs]: We can remove that, but the developers will have a lot of troubles when the method will throw in their tests, they will figure out why, then they will setup the context...

**Comment by brianlagunas** [comment]
> I already have this in my application

Great, that is where it belongs.

> If you continue your reasoning, you end up with the Prism is not needed at all - developers can write all the code by themselves

Technically they can now, but that's a lot of code to write.  There must be a balance between providing a helpful framework, and keeping clear and succinct API.  Adding that parameter is just bad API design.  I cannot accept this PR if that parameter is there.

File: Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs (hunk 2 above)
Comment (brianlagunas) [inline Source/Wpf/Prism.Wpf/Interactivity/InteractionRequest/InteractionRequest.cs]: That is the nature of testing async code that relies on a SyncContext, but we do not place code into production that is solely meant to support test libraries.  If the code cannot be tested as-is, it points to code smell.

**Comment by dansiegel** [comment]
@dvorn not true at all.

```cs
public Task<T> RaiseAsync(T context, bool executeSynchronously)
{
    if (executeSynchronously)
    {
        return CallbackHelper.AwaitCallbackResult<T>(callback => this.Raise(context, callback));
    }
    else
    {
        return this.RaiseImplAsync(context);
    }
}
```

becomes...

```cs
public Task<T> RaiseAsync(T context)
{
    return this.RaiseImplAsync(context);
}

public T Raise(T context)
{
    return CallbackHelper.AwaitCallbackResult<T>(callback => this.Raise(context, callback));
}
```

Having a boolean that determines whether or not a method runs Synchronously or not is just a bad pattern, particularly when the method name indicates that it should be an Asynchronous method.

**Comment by dvorn** [comment]
@dansiegel @brianlagunas  See sample code in related issue: https://github.com/PrismLibrary/Prism/issues/678#issuecomment-238692910

**Comment by dvorn** [comment]
@dansiegel 
```csharp
public T Raise(T context)
{
    return CallbackHelper.AwaitCallbackResult<T>(callback => this.Raise(context, callback));
}
```
You just do not understand.  "T Raise(T context)" already exists.  It is a base synchronous version.  AwaitCallbackResult returns not T but Task<T>.  And you call Raise() recursively.

**Comment by dansiegel** [comment]
@dvorn if you already have a Synchronous version I simply do not understand what your problem is or why you feel you need to wrap your return type in a Task. I'm sorry but based on what I see I would agree with @brianlagunas that this shouldn't be included as part of the Prism API.

**Comment by dvorn** [comment]
@brianlagunas @dansiegel As it turns out, I am kinda reinvented the wheel here with the `executeSynchronously` parameter.  This trick was suggested by **Stephen Toub** and elaborated on by **Stephen Cleary** in [Async Programming - Brownfield Async Development](https://msdn.microsoft.com/en-us/magazine/mt238404.aspx), section **The Flag Argument Hack**.
```csharp
public interface IDataService
{
  string Get(int id);
  Task<string> GetAsync(int id);
  Task<string> GetCoreAsync(int id, bool sync);
}
```
**StephenCleary**:

>  it’s still my favorite of the hacks.

Note that the code in Figure 11 in the article where the sync argument propagates up into the business logic is fully analogous to the sample I gave in https://github.com/PrismLibrary/Prism/issues/678#issuecomment-238692910

**Comment by brianlagunas** [comment]
I am removing RaiseAsync form the Prism API. See #858

**Pull request closed without merging**